src/durmod/kaldi-durmod.h

```diff
@@ -0,0 +1,266 @@
+#include "util/stl-utils.h"
+#include "fstext/deterministic-fst.h"
+
+namespace kaldi {
+  using nnet3::Nnet;
+  using nnet3::NnetExample;
+  using nnet3::NnetIo;
+  using nnet3::CachingOptimizingCompiler;
+  using nnet3::ComputationRequest;
+  using nnet3::NnetComputation;
+  using nnet3::NnetComputeOptions;
+  using nnet3::NnetComputer;
+
+struct PhoneDurationModelOptions {
+  int32 left_context, right_context;
+  int32 max_duration;
+
+  void Register(OptionsItf *po);
+
+  PhoneDurationModelOptions():
+    left_context(4),
+    right_context(2),
+    max_duration(60) {}
+};
+
+class PhoneDurationModel {
+  friend class PhoneDurationFeatureMaker;
+
+ public:
+  void Read(std::istream &is, bool binary);
+  void Write(std::ostream &os, bool binary) const;
+
+  PhoneDurationModel() {}
+  PhoneDurationModel(const PhoneDurationModelOptions &opts,
+                     const std::vector<std::vector<int32> > &roots,
+                     const std::vector<std::vector<int32> > &questions):
+    left_context_(opts.left_context),
+    right_context_(opts.right_context),
+    roots_(roots),
+    questions_(questions),
+    max_duration_(opts.max_duration) {}
+
+  inline int32 MaxDuration() const { return max_duration_; }
+  inline int32 RightContext() const { return right_context_; }
+  inline int32 LeftContext() const { return left_context_; }
+  inline int32 FullContextSize() const {
+    return left_context_ + right_context_ + 1;
+  }
+  std::string Info() const;
+
+ private:
+  int32 left_context_, right_context_;
+  std::vector<std::vector<int32> > roots_;
+  std::vector<std::vector<int32> > questions_;
+  int32 max_duration_;
+};
+
+class PhoneDurationFeatureMaker {
+ public:
+  explicit PhoneDurationFeatureMaker(const PhoneDurationModel &model);
+
+  void InitFeatureMaker(const PhoneDurationModel &model);
+
+  /// This method extracts features for a phone. The inputs are a
+  /// phone-duration context and an index into that context which indicates the
+  /// middle phone (middle in the sense of left/right context).
+  /// a phone-duration context is a sequence of <phone-ID, duration-in-frames>
+  /// pairs.
+  /// The extracted features consist of phone-IDs (which are represented in a
+  /// 1-of-n encoding, suitable for neural networks), phone durations, and
+  /// binary features determined from a set of questions
+  /// (i.e. extra_questions.int)
+  void MakeFeatureVector(
+                 const std::vector<std::pair<int32, int32> > &phone_dur_context,
+                 int phone_index,
+                 SparseVector<BaseFloat> *feat) const;
+
+
+  int32 FeatureDim() const { return feature_dim_; }
+  int32 NumBinaryFeatures() const { return num_binary_features_; }
+  int32 NumPhoneIdentities() const { return num_phone_identities_; }
+  int32 OutputDim() const { return max_duration_; }
+
+  std::string Info() const;
+
+ private:
+  unordered_map<int32, std::vector<int32> > binary_feats_;
+  unordered_map<int32, int32> phone_id_;
```

Comment:
So if I get it right, this means that we train a separate neural network  (of course implemented in the same topology of our main nnet) - with its input not including durations - to learn logprobs (computed using main neural network) of the input (middle) phones. However, the logprobs (i.e. the target of this separate nnet) will fluctuate a lot; both because the main network is being trained and more importantly because for any unique input (i.e. phone-context without durations) to this separate nnet, we will get different logprobs based on the duration of the middle phone. Isn't that going to cause a problem?